### PR TITLE
Implemented ability to search iTunes for a series and add to database

### PR DESCRIPTION
**Pull request by majora2007**
#### Database Migration
YES

#### Description
Added the ability to search iTunes for an artist and add an artist to the database. 

#### Issues Fixed or Closed by this PR

*

### CODE REVIEW

**Comment by TCBWZA** [comment]
I've scanned through the layout and think we need a couple of changes.

The current layout does not allow for Compilations (from my quick scan) where the track artist will be different from the Album Artist.

Also I think we need a table for reference where there is more than 1 Album Artist e.g. collaborations.

When we do an import we may not know what the artists values are. Also we need to allow for correction of different spelling for the same artist e.g. Björk vs Bjork.

**Comment by nbonfire** [comment]
Looks like you considered multi-disc albums in the album object, but there's no corresponding "disc number" in the track info. 

Also I would recommend adding all the fields that are stored in iTunes:
* a "media kind" (audiobook vs music vs comedy),
* bpm, 
* composer,
* genre,
* "sort as" tags for album, album artist, artist, track,  etc.

**Comment by majora2007** [comment]
Hey @TCBWZA, I believe this can allow for compilations. If instead of bundling multiple artists with a track, we instead bundle them with an Album, we should be able to accomplish the same thing. Remember, this tool is aimed at people who download whole albums for completing their libraries. Likely, they won't be downloading one-off songs. 

But I do agree, we should have a flag on the Album if it is a collaboration and link to a separate table. 

For Björk vs Bjork, wouldn't the artistSlug handle this? It would clean up and conform to a set standard. As when you import, we need to clean each title for the search. Then from that cleaned title, we can use it to search.

@nbonfire, genre can be added, but can you explain any value gain from having "sort as", composer, bpm, or "media kind"? When we search Indexers, they will unlikely have this information available.

**Comment by majora2007** [comment]
On second thought, I agree with you @TCBWZA. I misunderstood compilation tracks. If we have a compilationId in the Track table that links to a table "Compilations" that has the artistIds we should be good, right?

Also, for clean titles, I was looking at my library and I saw some titles where the software could never clean. See below:

image

Not sure how to handle this.

**Comment by majora2007** [comment]
Updated PR with changes for Compilations. Please review.

**Comment by TCBWZA** [comment]
I'm concerned that this will complicate the queries. Cartesian product for every artist involved in a single track. How do you search for collaborations and compilations?

File: src/NzbDrone.Core/Datastore/TableMapping.cs [hunk 1]
```diff
@@ -112,6 +112,9 @@ public static void Map()
                   .Relationship()
                   .HasOne(track => track.TrackFile, track => track.TrackFileId);
 
+            Mapper.Entity<Compilation>().RegisterModel("Compilation")
+                   .Relationships.AutoMapICollectionOrComplexProperties(); //TODO: Figure out how to map this Table
```
Comment (galli-leo) [inline src/NzbDrone.Core/Datastore/TableMapping.cs]: According to the Mapper docs (https://marrdatamapper.codeplex.com/wikipage?title=Mappings&referringTitle=Documentation), this should work:
```
Relationships.MapProperties()
.For("Artists") 
.LazyLoad((db, comp) => 
db.Query<Artist>().Where(art => art.CompilationId == comp.CompilationId).ToList())```

**Comment by nbonfire** [comment]
My thinking for all those other tags is for moving/renaming. Also I was under the impression that Lidarr was going to pull album/artist/track info from iTunes, not the indexer.

**Comment by majora2007** [comment]
I think we can instead keep them with the metadata rather than in the table
itself. That is how Sonarr does it.

**Comment by mattman86** [comment]
I thought the plan was iTunes as well.

**Comment by majora2007** [comment]
Yes, we are using iTunes. The way it works in my eyes is the following:

I import a file, the track gets put in the database and a MediaFile is associated with it. The MediaFile, like in Sonarr, will have all information about the track model. BPM, Codec, File Size, etc.

To generate the MediaFile, I will use MediaInfo/read the IDv3 tags. 

Hence there is no need to store extra information about the file in the database. 

Only issues I can see is the "clean title issue" aka UTF32 in our database or being able to call iTunes API with titles like †‡† and how we need to model Compilations, suggestions (with psuedocode) very welcome. Might be better to discuss on discord.

**Comment by majora2007** [comment]
@nbonfire are you interested in developing Lidarr?

**Comment by nbonfire** [comment]
Unfortunately I don't have the time to learn c# and contribute much

**Pull request merged**